### PR TITLE
[SPARK-5119] java.lang.ArrayIndexOutOfBoundsException on trying to train...

### PR DESCRIPTION
**Pull request by Lewuathe**
... decision tree model

Labels loaded from libsvm files are mapped to 0.0 if they are negative labels because they should be nonnegative value.

### CODE REVIEW

**Comment by srowen** [comment]
On the one hand, this changes behavior, and causes the result to not quite match what's in the input file. Some parts of MLlib handle -1 / +1 for binary classification, although generally 0 / 1 is expected. That said, this method already changes what it reads by modifying 1-based feature indices to 0-based.

FWIW I think this would be a good change since -1 / +1 is sometimes used in libsvm, and -1 does not always work within MLlib. Some docs and a test to accompany this would be appropriate. You might also search the code base for instances where this translation already happens and remove it if needed, but I didn't see such a thing.

(And the JIRA / PR title could be updated to reflect what the issue really is now)

**Comment by mengxr** [comment]
This is a silent behavior change, which may cause unexpected errors on user side. Does the data validators in each algorithm check label range?

**Comment by Lewuathe** [comment]
@mengxr Thank you. I'll check it.

**Comment by Lewuathe** [comment]
@srowen  @mengxr There seems to be no checking labels logics on each algorithms when importing `LabeledPoint`. But these are only check for first access to `LabeledPoint`. If anyone make checking code to `LabeledPoint` as client or jobs for spark, I found those will broken. I'm sorry but eventually it is better to throw exceptions which means `Impurity` does not support negative labels. What do you think about it?

**Comment by jkbradley** [comment]
My 2 cents: I think the right places to do the transformation and check are:
* Transforming -1/+1 to 0/1: This should be done by the loadLibSVMFile methods automatically.
* Checking for 0-based indexed labels: This should be done by each algorithm when it is called.  That way, behavior is not changed.

Note for the spark.ml package: The behavior can be improved as we move to SchemaRDDs with metadata.  A label column should have metadata about the labels (numClasses, plus a dictionary if they are mapped to 0-based indices).  Data loading methods should add this metadata automatically.  Algorithms should see if the metadata exists and call a column indexing method if not.

**Comment by Lewuathe** [comment]
@jkbradley Is transforming -1/+1 to 0/1 behaviour change from the viewpoint of each algorithms? You mean users should recognize the fact that `loadLibSVMFile` maps negative labels to positive ones automatically?

**Comment by jkbradley** [comment]
@Lewuathe  You're right; it is an API change for either algorithm.  I was arguing for an API change in data loading.  However, thinking more about it, I think you're right to make the API change happen in DecisionTree itself.  (I've written out my thoughts in more detail below.)  Here are some questions which come up though:
* Do we only re-label -1,+1 to 0,1?  Or do we re-label any set of values to 0,1,2,...?
* Does predict() return the original labels or the re-labeled ones?
* Should we make this standard across all algorithms (changing even more APIs)?

CC: @mengxr  What do you think?

My various thoughts:

*In the spark.mllib API:*
* If DecisionTree relabels -1 to 0, then we must do 1 of the following for test-time prediction (and neither option is great):
  * Prediction outputs original labels.
    * This would require saving a label dictionary/index mapping between original labels and new labels.
    * Con: Some methods for computing test error will no longer work (e.g., sum_i (prediction_i - label_i)^2).
  * Prediction outputs 0-based indexed labels.
    * Con: Users will get confused when they test on held-out data which has not been relabeled.
* LibSVM does not impose a standard for labels, but it really should.

*In the spark.ml API:*
Re-indexing should happen in DecisionTree.  This will be easier to do since SchemaRDD metadata can store the label dictionary/index.

**Comment by mengxr** [comment]
LIBSVM is also used for regression data. I don't think we can transform label values automatically for users inside `loadLibSVMFile`. There used to be a `multiclass` flag, but we decided to deprecate its usage and load the raw data from LIBSVM files.

The label set should be validated inside each algorithm, e.g., 

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/util/DataValidators.scala

@jkbradley It would be easier for the algorithms takes 0,1,2,... and make predictions in the same domain. For test data, users should apply the same sequence of transformations that were used to produce the training dataset. We can embed the mapping from raw labels to indexed labels in the metadata.

**Comment by jkbradley** [comment]
@mengxr  I agree that setup is simplest, in that it makes users consciously handle all re-labeling explicitly.  I'm OK with it if we add data validation at the start of each algorithm (or in a try-except block which then does data validation when an exception is thrown).

**Comment by Lewuathe** [comment]
@mengxr @jkbradley You mean `LabeledPoint` should keep mapping meta data which transform raw data to re-labeled points? Or do you think user should prepare their validators and transformers when they use each algorithms?

**Comment by jkbradley** [comment]
@Lewuathe  I believe @mengxr meant that users should prepare transformers whenever they use algorithms, as follows:
* User loads data
* User maps labels to 0,1,2,... to create newData
* User passes newData to learning algorithm.
* Learning algorithm validates newData, throwing an exception if the labels are not in 0,1,2,...
* User calls model.predict to get predicted labels in 0,1,2,...
* User can transform those predicted labels back to the original labels if needed.

I'm fine with this.  We'll just have to make it as easy as possible to do this mapping.  The main complication I see in the above workflow is in the pipelines API: A user would want to add a Transformer to the Pipeline to map labels to 0,1,2..., and the user would need another Transformer to map the predictions back to the original set of labels.  But those 2 Transformers should be linked somehow so that fitting the first (to compute the label index/dictionary) also fits the second.  @mengxr  Any thoughts on this?  It sounds doable, but it adds a new sort of pipeline feature (linked PipelineStages).

**Comment by mengxr** [comment]
@Lewuathe I'm suggesting the following:

1) loadLibSVMFile should not transform the input data.
2) Users should map the labels to 0, 1, 2, 3, ...
  a) using ad-hoc transformations,
  b) or using a `LabelIndexer` (to be added) in the new pipeline API, where metadata will be added to the output column of `LabelIndexer`
3) Algorithms should validate the label values and throw errors if they are out of range

**Comment by Lewuathe** [comment]
@mengxr @jkbradley Thank you for suggestive advices. Anyway `loadLibSVMFile` should not be changed and in this case the problem is occurred by negative label in `Impurity`. So I update `Impurity` to throw exception.

**Comment by jkbradley** [comment]
@Lewuathe  Sounds good.  Thanks for going through this discussion!  Looking at the latest updates, it looks like the 2 Impurity tests are very similar.  Would you be able to put them in one ImpuritySuite file and combine the tests to shorten the code (e.g., by using a type parameter for the Impurity type)?

File: mllib/src/main/scala/org/apache/spark/mllib/tree/impurity/Entropy.scala [hunk 1]
```diff
@@ -94,6 +94,9 @@ private[tree] class EntropyAggregator(numClasses: Int)
       throw new IllegalArgumentException(s"EntropyAggregator given label $label" +
         s" but requires label < numClasses (= $statsSize).")
     }
+    if (label < 0) {
+      throw new IllegalArgumentException(s"Entropy does not support negative labels")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/impurity/Entropy.scala]: Shall we put `label` value to the error message? Add `.` to the end of the sentence.

File: mllib/src/test/scala/org/apache/spark/mllib/tree/EntropySuite.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.tree
+
+import org.apache.spark.mllib.tree.impurity._
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/EntropySuite.scala]: Try to be specific if only few classes are really used.

File: mllib/src/test/scala/org/apache/spark/mllib/tree/EntropySuite.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.scalatest.FunSuite
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/EntropySuite.scala]: organize imports (https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports)

File: mllib/src/test/scala/org/apache/spark/mllib/tree/EntropySuite.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Test suite for [[org.apache.spark.mllib.tree.impurity.Entropy]]
+ */
+class EntropySuite extends FunSuite with MLlibTestSparkContext{
+  test("Gini impurity does not support negative labels") {
+    val entropy = new EntropyAggregator(2)
+    try {
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/EntropySuite.scala]: This is more common in the Spark codebase:

~~~
intercept[IllegalArgumentException] {
  entropy.update(Array(0.0, 1.0, 2.0), 0, -1, 0.0)
}
~~~

File: mllib/src/test/scala/org/apache/spark/mllib/tree/GiniImpuritySuite.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.tree
+
+import org.apache.spark.mllib.tree.impurity._
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.scalatest.FunSuite
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/GiniImpuritySuite.scala]: organize imports

File: mllib/src/test/scala/org/apache/spark/mllib/tree/ImpuritySuite.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.tree
+
+import org.scalatest.FunSuite
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/ImpuritySuite.scala]: There should be an empty line after this import. Please check the style guide.

Comment (Lewuathe) [inline mllib/src/test/scala/org/apache/spark/mllib/tree/ImpuritySuite.scala]: I'm sorry for missing it. I fixed it.

**Comment by mengxr** [comment]
LGTM pending Jenkins.

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**